src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+        end_time = cast(datetime, min_max.max())
+        # Add one minute to end_time, because time intervals are right
+        # open and one minute is the smallest interval allowed.
+        stop_time = end_time + timedelta(minutes=1)
+        # Round down to the nearest minute.
+        start = start_time.replace(second=0, microsecond=0)
+        stop = stop_time.replace(second=0, microsecond=0)
+        return TimeRange(start, stop)
+
+    def __iter__(self) -> Iterator[Event]:
+        for i, event in self.iterrows():
+            yield Event(
+                event,
+                event_id=EventId(i, self._self_role),
+                _model=self._self_model,
+            )
+
+    @overload
+    def __getitem__(self, key: ColumnKey) -> "pd.Series[Any]":
+        ...
+
+    @overload
+    def __getitem__(self, key: List[RowId]) -> "Events":
+        ...
+
+    def __getitem__(self, key: Any) -> Any:
+        if isinstance(key, list):
+            return Events(
+                self.loc[key],
+                role=self._self_role,
+                _model=self._self_model,
+            )
+        return super().__getitem__(key)
+
+
+class Dataset(Events):
+    """pd.DataFrame wrapped with extra functions and metadata."""
+
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        /,
+        name: str,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(df, **kwargs)
+        self._self_name = name
+
+    @property
+    def name(self) -> str:
+        return self._self_name
+
+    @property
+    def role(self) -> DatasetRole:
+        return self._self_role
+
+    @overload
+    def __getitem__(self, key: ColumnKey) -> "pd.Series[Any]":
+        ...
+
+    @overload
+    def __getitem__(self, key: List[RowId]) -> Events:
+        ...
+
+    def __getitem__(self, key: Any) -> Any:
+        if isinstance(key, list):
+            return Events(
+                self.iloc[key].set_axis(key),
+                role=self._self_role,
+                _model=self._self_model,
+            )
+        return super().__getitem__(key)
+
+
+@dataclass(frozen=True, repr=False)
+class Model:
+    """A Model consists of a set of dataframes and a set of Dimensions.
+    Dimensions are objects that extracts columns from dataframes. Not all
+    dataframe columns are Dimensions. Some are ancillary, e.g. a column of
+    URLs. All Dimensions have a primary column and its name is the same as
+    the Dimension's. Some Dimensions may have more than one column, but only
+    one is primary. When a Dimension tries to extract a non-existent column
+    from a dataframe, a default (or fallback) column will be returned, e.g.
+    a column of NaNs.
+    """
+
+    _datasets: Dict[DatasetRole, Dataset]
+    _dimensions: Dict[Name, Dimension]
+    _dim_roles: Dict[DimensionRole, List[Name]]
+    _original_columns: Dict[DatasetRole, pd.Index]
+    # now is used as the substitute for dataframes without timestamps.
+    _now: datetime
```

Comment:
Why does the model need to store this? just can be scoped to the timestamp generation?